src/vmm/src/vstate/vcpu/x86_64.rs

```diff
@@ -324,42 +324,76 @@ impl KvmVcpu {
     ///
     /// # Arguments
     ///
-    /// * `msr_index_list`: List of MSR indices.
+    /// * `msr_index_iter`: Iterator over MSR indices.
     ///
     /// # Errors
     ///
     /// * When [`kvm_bindings::Msrs::new`] returns errors.
     /// * When [`kvm_ioctls::VcpuFd::get_msrs`] returns errors.
     /// * When the return value of [`kvm_ioctls::VcpuFd::get_msrs`] (the number of entries that
     ///   could be gotten) is less than expected.
-    fn get_msr_chunks(&self, msr_index_list: &[u32]) -> Result<Vec<Msrs>, KvmVcpuError> {
-        let mut msr_chunks: Vec<Msrs> = Vec::new();
-
-        for msr_index_chunk in msr_index_list.chunks(KVM_MAX_MSR_ENTRIES) {
-            let mut msrs = Msrs::new(msr_index_chunk.len())?;
-            let msr_entries = msrs.as_mut_slice();
-            assert_eq!(msr_index_chunk.len(), msr_entries.len());
-            for (pos, index) in msr_index_chunk.iter().enumerate() {
-                msr_entries[pos].index = *index;
-            }
-
-            let expected_nmsrs = msrs.as_slice().len();
-            let nmsrs = self
-                .fd
-                .get_msrs(&mut msrs)
-                .map_err(KvmVcpuError::VcpuGetMsrs)?;
-            if nmsrs != expected_nmsrs {
-                return Err(KvmVcpuError::VcpuGetMsr(msr_index_chunk[nmsrs]));
-            }
-
-            msr_chunks.push(msrs);
+    fn get_msr_chunks(
+        &self,
+        mut msr_index_iter: impl ExactSizeIterator<Item = u32>,
+    ) -> Result<Vec<Msrs>, KvmVcpuError> {
+        let num_chunks = msr_index_iter.len() / KVM_MAX_MSR_ENTRIES;
+        let excess = msr_index_iter.len() % KVM_MAX_MSR_ENTRIES;
+
+        let mut msr_chunks: Vec<Msrs> = Vec::with_capacity(num_chunks);
+
+        for _ in 0..num_chunks {
+            let chunk = self.get_msr_chunk(&mut msr_index_iter, KVM_MAX_MSR_ENTRIES)?;
+            msr_chunks.push(chunk);
+        }
+        if excess != 0 {
+            let chunk = self.get_msr_chunk(&mut msr_index_iter, excess)?;
+            msr_chunks.push(chunk);
         }
```

Comment:
The `num_chunks` is not the correct length if `excess != 0`. How about this?
```suggestion
        let num_chunks = (msr_index_iter.len() + KVM_MAX_MSR_ENTRIES - 1) / KVM_MAX_MSR_ENTRIES;
        let mut msr_chunks: Vec<Msrs> = Vec::with_capacity(num_chunks);
        for _ in 0..num_chunks - 1 {
            let chunk = self.get_msr_chunk(&mut msr_index_iter, KVM_MAX_MSR_ENTRIES)?;
            msr_chunks.push(chunk);
        }
        let last = msr_index_iter.len() - KVM_MAX_MSR_ENTRIES * (num_chunks - 1);
        let chunk = self.get_msr_chunk(&mut msr_index_iter, last)?;
        msr_chunks.push(chunk);
```

Reply:
As far as my math goes, both approaches are correct. 
- In case the number of indexes is not divisible by `KVM_MAX_MSR_ENTRIES` both approaches do get first `n` full chunks in the first loop and then get last chunk as a second step.
- In case where  the number of indexes is divisible by `KVM_MAX_MSR_ENTRIES` my approach gets all chunks in the first loop and `excess` is 0, last chunk logic is ignored. In your case it will do `n-1` iterations in the first loop and will get last full chunk in the last part.

IMO my approach is simpler just because there is less math.

Reply:
I wish I could write a unit test for this, but the amount of msrs returned by `get_msrs_to_dump` or `vcpu.msrs_to_save` is smaller than `KVM_MAX_MSR_ENTRIES`

Reply:
What I meant is its capacity (`Vec::with_capacity(num_chunks);`) is not correct when the number of indices is not divisible by `KVM_MAX_MSR_ENTRIES`. The last `push()` requires it to gain another space.

Reply:
oh, yeah, I knew about this, but didn't give it a lot of attention. I guess I can just do `Vec::with_capacity(num_chunks + 1);` to be sure it has enough space.

Reply:
I'd like to hear @roypat 's opinion for this.

Reply:
How about just something like (pseudo-code)

```rs
self.msr_modifiers
     .chunks(KVM_MAX_MSR_ENTRIES)
     .map(|chunk| chunk.iter().map(|modifier| modifier.addr))
     .map(|chunk| self.get_msr_chunk(chunk))  // we have an exactsize iterator here, so can drop the "how many" parameter
     .collect::<Vec<_>>();
```

Reply:
unfortunately `get_msr_chunks` is used in `get_msrs`, so it needs to accept something that contains indexes.

Reply:
Alright, let's use https://doc.rust-lang.org/std/primitive.u32.html#method.div_ceil then :)
```rs
    fn get_msr_chunks(
        &self,
        mut msr_index_iter: impl ExactSizeIterator<Item = u32>,
    ) -> Result<Vec<Msrs>, KvmVcpuError> {
        let num_chunks = msr_index_iter.len().div_ceil(KVM_MAX_MSR_ENTRIES);

        // Allocate 1 more element in case excess is not 0
        let mut msr_chunks: Vec<Msrs> = Vec::with_capacity(num_chunks);

        for _ in 0..num_chunks {
            let len = msr_index_iter.len();
            let chunk = self.get_msr_chunk(&mut msr_index_iter, KVM_MAX_MSR_ENTRIES.min(len))?;
            msr_chunks.push(chunk);
        }

        Self::fix_zero_tsc_deadline_msr(&mut msr_chunks);

        Ok(msr_chunks)
    }
```

Reply:
done